### ISSUE TITLE
`metadata.configurations` got lost when decoding `buildpack.toml`

### ISSUE THREAD

**Issue by c0d1ngm0nk3y** [buildpacks/libcnb#299]
When creating the `DetectContext`, the `buildpack.toml` is decoded [here](https://github.com/buildpacks/libcnb/blob/f85e7067d6dd09793dbf1cafc7ddae523edffd9f/detect.go#L119) and stored into the [`Buildpack`](https://github.com/buildpacks/libcnb/blob/f85e7067d6dd09793dbf1cafc7ddae523edffd9f/buildpack.go#L115) struct.

```go
type Buildpack struct {
	// API is the api version expected by the buildpack.
	API string `toml:"api"`

	// Info is information about the buildpack.
	Info BuildpackInfo `toml:"buildpack"`

	// Path is the path to the buildpack.
	Path string `toml:"-"`

	// Deprecated: Stacks is the collection of stacks supported by the buildpack.
	Stacks []BuildpackStack `toml:"stacks"`

	// Targets is the collection of targets supported by the buildpack.
	Targets []Target `toml:"targets"`

	// Metadata is arbitrary metadata attached to the buildpack.
	Metadata map[string]interface{} `toml:"metadata"`
}
```

But when having a `buildpack.toml` containing `metadata.configurations` like [here](https://github.com/paketo-buildpacks/bellsoft-liberica/blob/0e3ff85ba2757d027a78ee46e558dfdf5b12aeab/buildpack.toml#L34), the configurations are not unmarhalled and therefor they are missing in the context.

This is due to

```go
	// Metadata is arbitrary metadata attached to the buildpack.
	Metadata map[string]interface{} `toml:"metadata"`
```

which works fine for simple [metadata](https://github.com/paketo-buildpacks/bellsoft-liberica/blob/0e3ff85ba2757d027a78ee46e558dfdf5b12aeab/buildpack.toml#L30) but not for the `configurations` objects.

**Comment by dmikusa** [comment]
[Metadata in buildpack.toml](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml) is defined by the spec as "buildpack-specific data". That means the buildpack decides what goes in there.

That makes it hard to do more than just `map[string]interface{}` here because from libcnb perspective it doesn't know what will be in the metadata. To be more strongly types, we'd either need to have something added to the spec, like have `metadata.configurations` be part of the spec, or we'd need to use generic types. I did a spike on generic types and it did help in some ways but it introduced other problems and some complexity with having types all over the place so I retracted that PR.

I suppose maybe we could introduce some helper methods for parsing out values from metadata. It isn't hard to parse out but it is tedious for sure. I haven't tried it but maybe there's something we could do with reflection to take in a struct, list the fields on the struct, look up those values safely in metadata, populate the struct, and return it. I haven't tried so not 100% sure it would possible, or how ergonomic you could make the method, but maybe something to investigate.

@c0d1ngm0nk3y Did you have something particular in mind here? A change you'd like to see made?

**Comment by c0d1ngm0nk3y** [comment]
@dmikusa Thanks for looking into this. After some more investigation, it looks like it works in general like expected. I guess there is no need to change anything right now. Sorry for the confusion.
